[BUG] Unsatisfiable error in stateful testing if all operations have inbound links

**Describe the bug**
Schemathesis stateful testing raises `hypothesis.errors.Unsatisfiable` when all operations have inbound links. In this case, all rules of the underlying state machine require some input from existing bundles, that are empty and can't be filled because they require some input to be generated. Recursion! :)

**To Reproduce**
1. Run the following test with `pytest`:

```python
SCHEMA = ...  # see below
schema = schemathesis.from_dict(SCHEMA, base_url="http://0.0.0.0:8081")

TestCase = schema.as_state_machine().TestCase
```

2. See 

```
E   hypothesis.errors.Unsatisfiable: Unable to satisfy assumptions of hypothesis run_state_machine.
```

API schema:

```json
{
    "openapi": "3.0.2",
    "info": {"title": "Test", "description": "Test", "version": "0.1.0"},
    "paths": {
        "/count": {
            "post": {
                "operationId": "counter",
                "requestBody": {
                    "content": {
                        "application/json": {
                            "schema": {
                                "type": "integer",
                                "minimum": 0,
                                "maximum": 5
                            }
                        }
                    },
                    "required": true
                },
                "responses": {
                    "200": {
                        "description": "OK",
                        "links": {
                            "CountMore": {
                                "operationId": "counter",
                                "requestBody": "$response.body#/value"
                            }
                        }
                    }
                }
            }
        }
    },
}
```

**Expected behavior**
Schemathesis won't fail with such error and will generate data from the operation definition

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -36,6 +36,7 @@ Changelog
 - Internal error if filling missing explicit examples led to ``Unsatisfiable`` errors. `#904`_
 - Do not suggest to disable schema validation if it is already disabled. `#914`_
 - Skip explicit examples generation if this phase is disabled via config. `#905`_
+- ``Unsatisfiable`` error in stateful testing caused by all API operations having inbound links. `#965`_, `#822`_
 
 **Removed**
 
@@ -1620,6 +1621,7 @@ Deprecated
 .. _0.3.0: https://github.com/schemathesis/schemathesis/compare/v0.2.0...v0.3.0
 .. _0.2.0: https://github.com/schemathesis/schemathesis/compare/v0.1.0...v0.2.0
 
+.. _#965: https://github.com/schemathesis/schemathesis/issues/965
 .. _#951: https://github.com/schemathesis/schemathesis/issues/951
 .. _#945: https://github.com/schemathesis/schemathesis/issues/945
 .. _#941: https://github.com/schemathesis/schemathesis/issues/941
@@ -1656,6 +1658,7 @@ Deprecated
 .. _#832: https://github.com/schemathesis/schemathesis/issues/832
 .. _#830: https://github.com/schemathesis/schemathesis/issues/830
 .. _#824: https://github.com/schemathesis/schemathesis/issues/824
+.. _#822: https://github.com/schemathesis/schemathesis/issues/822
 .. _#816: https://github.com/schemathesis/schemathesis/issues/816
 .. _#814: https://github.com/schemathesis/schemathesis/issues/814
 .. _#812: https://github.com/schemathesis/schemathesis/issues/812

--- a/src/schemathesis/specs/openapi/stateful/__init__.py
+++ b/src/schemathesis/specs/openapi/stateful/__init__.py
@@ -1,7 +1,7 @@
 import functools
 import operator
 from collections import defaultdict
-from typing import TYPE_CHECKING, Any, Dict, List, Tuple, Type
+from typing import TYPE_CHECKING, Any, Dict, Generator, List, Tuple, Type
 
 from hypothesis.stateful import Bundle, Rule, rule
 from hypothesis.strategies import SearchStrategy, none
@@ -61,21 +61,22 @@ def make_all_rules(
 ) -> Dict[str, Rule]:
     """Create rules for all endpoints, based on the provided connections."""
     return {
-        f"rule {endpoint.verbose_name}": make_rule(
-            endpoint, bundles[endpoint.path][endpoint.method.upper()], connections
-        )
+        f"rule {endpoint.verbose_name} {idx}": new
         for endpoint in schema.get_all_endpoints()
+        for idx, new in enumerate(make_rules(endpoint, bundles[endpoint.path][endpoint.method.upper()], connections))
     }
 
 
-def make_rule(endpoint: "Endpoint", bundle: Bundle, connections: EndpointConnections) -> Rule:
+def make_rules(endpoint: "Endpoint", bundle: Bundle, connections: EndpointConnections) -> Generator[Rule, None, None]:
     """Create a rule for an endpoint."""
+
+    def _make_rule(previous: SearchStrategy) -> Rule:
+        return rule(target=bundle, previous=previous, case=endpoint.as_strategy())(APIStateMachine.step)  # type: ignore
+
     previous_strategies = connections.get(endpoint.verbose_name)
     if previous_strategies is not None:
-        previous = _combine_strategies(previous_strategies)
-    else:
-        previous = none()
-    return rule(target=bundle, previous=previous, case=endpoint.as_strategy())(APIStateMachine.step)  # type: ignore
+        yield _make_rule(_combine_strategies(previous_strategies))
+    yield _make_rule(none())
 
 
 def _combine_strategies(strategies: List[SearchStrategy]) -> SearchStrategy:

--- a/test/apps/_aiohttp/__init__.py
+++ b/test/apps/_aiohttp/__init__.py
@@ -58,7 +58,6 @@ def create_openapi_app(
         + [web.route(item.value[0], item.value[1], wrapper(item.name)) for item in Endpoint if item.name != "all"]
     )
     app["users"] = {}
-    app["requests_history"] = defaultdict(list)
     app["incoming_requests"] = incoming_requests
     app["schema_requests"] = schema_requests
     app["config"] = {"should_fail": True, "schema_data": make_openapi_schema(endpoints, version)}
@@ -72,7 +71,6 @@ def reset_app(
 ) -> None:
     """Clean up all internal containers of the application and resets its config."""
     app["users"].clear()
-    app["requests_history"].clear()
     app["incoming_requests"][:] = []
     app["schema_requests"][:] = []
     app["config"].update({"should_fail": True, "schema_data": make_openapi_schema(endpoints, version)})

--- a/test/apps/_aiohttp/handlers.py
+++ b/test/apps/_aiohttp/handlers.py
@@ -3,6 +3,7 @@ import cgi
 import csv
 import io
 from typing import Dict
+from uuid import uuid4
 
 import jsonschema
 from aiohttp import web
@@ -11,15 +12,6 @@ try:
     from ..utils import PAYLOAD_VALIDATOR
 except (ImportError, ValueError):
     from utils import PAYLOAD_VALIDATOR
-
-
-def get_integer_parameter(request: web.Request, name: str) -> int:
-    try:
-        return int(request.match_info[name])
-    except KeyError:
-        raise web.HTTPBadRequest(text=f'{{"detail": "Missing `{name}`"}}')
-    except ValueError:
-        raise web.HTTPBadRequest(text=f'{{"detail": "Invalid `{name}`"}}')
 
 
 async def expect_content_type(request: web.Request, value: str):
@@ -190,40 +182,44 @@ async def form(request: web.Request) -> web.Response:
 
 async def create_user(request: web.Request) -> web.Response:
     data = await request.json()
-    if "username" not in data:
-        raise web.HTTPBadRequest(text='{"detail": "Missing `username`"}')
-    if not isinstance(data["username"], str):
-        raise web.HTTPBadRequest(text='{"detail": "Invalid `username`"}')
-    user_id = len(request.app["users"]) + 1
+    for field in ("first_name", "last_name"):
+        if field not in data:
+            raise web.HTTPBadRequest(text=f'{{"detail": "Missing `{field}`"}}')
+        if not isinstance(data[field], str):
+            raise web.HTTPBadRequest(text=f'{{"detail": "Invalid `{field}`"}}')
+    user_id = str(uuid4())
     request.app["users"][user_id] = {**data, "id": user_id}
-    request.app["requests_history"][user_id].append("POST")
     return web.json_response({"id": user_id}, status=201)
 
 
+def get_user_id(request: web.Request) -> str:
+    try:
+        return request.match_info["user_id"]
+    except KeyError:
+        raise web.HTTPBadRequest(text='{"detail": "Missing `user_id`"}')
+
+
 async def get_user(request: web.Request) -> web.Response:
-    user_id = get_integer_parameter(request, "user_id")
+    user_id = get_user_id(request)
     try:
         user = request.app["users"][user_id]
-        request.app["requests_history"][user_id].append("GET")
-        return web.json_response(user)
+        # The full name is done specifically via concatenation to trigger a bug when the last name is `None`
+        full_name = user["first_name"] + " " + user["last_name"]
+        return web.json_response({"id": user["id"], "full_name": full_name})
     except KeyError:
         return web.json_response({"message": "Not found"}, status=404)
 
 
 async def update_user(request: web.Request) -> web.Response:
-    user_id = get_integer_parameter(request, "user_id")
+    user_id = get_user_id(request)
     try:
         user = request.app["users"][user_id]
-        history = request.app["requests_history"][user_id]
-        history.append("PATCH")
-        if history == ["POST", "GET", "PATCH", "GET", "PATCH"]:
-            raise web.HTTPInternalServerError(text="We got a problem!")
         data = await request.json()
-        if "username" not in data:
-            raise web.HTTPBadRequest(text='{"detail": "Missing `username`"}')
-        if not isinstance(data["username"], str):
-            raise web.HTTPBadRequest(text='{"detail": "Invalid `username`"}')
-        user["username"] = data["username"]
+        for field in ("first_name", "last_name"):
+            if field not in data:
+                raise web.HTTPBadRequest(text=f'{{"detail": "Missing `{field}`"}}')
+            # Here we don't check the input value type to emulate a bug in another operation
+            user[field] = data[field]
         return web.json_response(user)
     except KeyError:
         return web.json_response({"message": "Not found"}, status=404)

--- a/test/apps/_fastapi/__init__.py
+++ b/test/apps/_fastapi/__init__.py
@@ -1,4 +1,5 @@
-from collections import defaultdict
+from typing import Optional
+from uuid import uuid4
 
 from fastapi import FastAPI, HTTPException, Query
 from pydantic import BaseModel, Field
@@ -7,7 +8,16 @@ from ..utils import OpenAPIVersion
 
 
 class User(BaseModel):
-    username: str = Field(min_length=3)
+    first_name: str = Field(min_length=3)
+    last_name: str = Field(min_length=3)
+
+    class Config:
+        extra = "forbid"
+
+
+class BuggyUser(BaseModel):
+    first_name: str = Field(min_length=3)
+    last_name: Optional[str] = Field(min_length=3, nullable=True)
 
     class Config:
         extra = "forbid"
@@ -25,7 +35,6 @@ def create_app(endpoints=("root",), version=OpenAPIVersion("3.0")):
         raise ValueError("FastAPI supports only Open API 3.0")
     app = FastAPI()
     users = {}
-    history = defaultdict(list)
 
     if "root" in endpoints:
 
@@ -37,32 +46,35 @@ def create_app(endpoints=("root",), version=OpenAPIVersion("3.0")):
 
         @app.post("/users/", status_code=201)
         def create_user(user: User):
-            user_id = len(users) + 1
+            user_id = str(uuid4())
             users[user_id] = {**user.dict(), "id": user_id}
-            history[user_id].append("POST")
             return {"id": user_id}
 
     if "get_user" in endpoints:
 
         @app.get("/users/{user_id}", responses={404: {"model": Message}})
-        def get_user(user_id: int, uid: int = Query(...), code: int = Query(...)):
+        def get_user(user_id: str, uid: str = Query(...), code: int = Query(...)):
             try:
                 user = users[user_id]
-                history[user_id].append("GET")
-                return user
+                # The full name is done specifically via concatenation to trigger a bug when the last name is `None`
+                try:
+                    full_name = user["first_name"] + " " + user["last_name"]
+                except TypeError:
+                    # We test it via out ASGI integration and `TypeError` will be propagated otherwise.
+                    # To keep the same behavior across all test server implementations we reraise it as a server error
+                    raise HTTPException(status_code=500, detail="We got a problem!")
+                return {"id": user["id"], "full_name": full_name}
             except KeyError:
                 raise HTTPException(status_code=404, detail="Not found")
 
     if "update_user" in endpoints:
 
         @app.patch("/users/{user_id}", responses={404: {"model": Message}})
-        def update_user(user_id: int, update: User, common: int = Query(...)):
+        def update_user(user_id: str, update: BuggyUser, common: int = Query(...)):
             try:
                 user = users[user_id]
-                history[user_id].append("PATCH")
-                if history[user_id] == ["POST", "GET", "PATCH", "GET", "PATCH"]:
-                    raise HTTPException(status_code=500)
-                user["username"] = update.username
+                for field in ("first_name", "last_name"):
+                    user[field] = getattr(update, field)
                 return user
             except KeyError:
                 raise HTTPException(status_code=404, detail="Not found")

--- a/test/apps/_flask/__init__.py
+++ b/test/apps/_flask/__init__.py
@@ -1,9 +1,9 @@
 import cgi
 import csv
 import json
-from collections import defaultdict
 from time import sleep
 from typing import Tuple
+from uuid import uuid4
 
 import jsonschema
 import yaml
@@ -33,7 +33,6 @@ def create_openapi_app(
     app.config["schema_requests"] = []
     app.config["internal_exception"] = False
     app.config["users"] = {}
-    app.config["requests_history"] = defaultdict(list)
 
     @app.before_request
     def store_request():
@@ -199,38 +198,35 @@ def create_openapi_app(
     @app.route("/api/users/", methods=["POST"])
     def create_user():
         data = request.json
-        if "username" not in data:
-            return jsonify({"detail": "Missing `username`"}), 400
-        if not isinstance(data["username"], str):
-            return jsonify({"detail": "Invalid `username`"}), 400
-        user_id = len(app.config["users"]) + 1
+        for field in ("first_name", "last_name"):
+            if field not in data:
+                return jsonify({"detail": f"Missing `{field}`"}), 400
+            if not isinstance(data[field], str):
+                return jsonify({"detail": f"Invalid `{field}`"}), 400
+        user_id = str(uuid4())
         app.config["users"][user_id] = {**data, "id": user_id}
-        app.config["requests_history"][user_id].append("POST")
         return jsonify({"id": user_id}), 201
 
-    @app.route("/api/users/<int:user_id>", methods=["GET"])
+    @app.route("/api/users/<user_id>", methods=["GET"])
     def get_user(user_id):
         try:
             user = app.config["users"][user_id]
-            app.config["requests_history"][user_id].append("GET")
-            return jsonify(user)
+            # The full name is done specifically via concatenation to trigger a bug when the last name is `None`
+            full_name = user["first_name"] + " " + user["last_name"]
+            return jsonify({"id": user["id"], "full_name": full_name})
         except KeyError:
             return jsonify({"message": "Not found"}), 404
 
-    @app.route("/api/users/<int:user_id>", methods=["PATCH"])
+    @app.route("/api/users/<user_id>", methods=["PATCH"])
     def update_user(user_id):
         try:
             user = app.config["users"][user_id]
-            history = app.config["requests_history"][user_id]
-            history.append("PATCH")
-            if history == ["POST", "GET", "PATCH", "GET", "PATCH"]:
-                raise InternalServerError("We got a problem!")
             data = request.json
-            if "username" not in data:
-                return jsonify({"detail": "Missing `username`"}), 400
-            if not isinstance(data["username"], str):
-                return jsonify({"detail": "Invalid `username`"}), 400
-            user["username"] = data["username"]
+            for field in ("first_name", "last_name"):
+                if field not in data:
+                    return jsonify({"detail": f"Missing `{field}`"}), 400
+                # Here we don't check the input value type to emulate a bug in another operation
+                user[field] = data[field]
             return jsonify(user)
         except KeyError:
             return jsonify({"message": "Not found"}), 404

--- a/test/apps/utils.py
+++ b/test/apps/utils.py
@@ -233,8 +233,11 @@ def _make_openapi_2_schema(endpoints: Tuple[str, ...]) -> Dict:
                         "required": True,
                         "schema": {
                             "type": "object",
-                            "properties": {"username": {"type": "string", "minLength": 3}},
-                            "required": ["username"],
+                            "properties": {
+                                "first_name": {"type": "string", "minLength": 3},
+                                "last_name": {"type": "string", "minLength": 3},
+                            },
+                            "required": ["first_name", "last_name"],
                             "additionalProperties": False,
                         },
                     }
@@ -246,7 +249,6 @@ def _make_openapi_2_schema(endpoints: Tuple[str, ...]) -> Dict:
                 {
                     "operationId": "updateUser",
                     "parameters": {"user_id": "$response.body#/id"},
-                    "requestBody": {"username": "foo"},
                 },
             )
             template["x-components"]["responses"] = {
@@ -266,12 +268,12 @@ def _make_openapi_2_schema(endpoints: Tuple[str, ...]) -> Dict:
             }
         elif endpoint == "get_user":
             parent = template["paths"].setdefault(path, {})
-            parent["parameters"] = [{"in": "path", "name": "user_id", "required": True, "type": "integer"}]
+            parent["parameters"] = [{"in": "path", "name": "user_id", "required": True, "type": "string"}]
             schema = {
                 "operationId": "getUser",
                 "parameters": [
                     {"in": "query", "name": "code", "required": True, "type": "integer"},
-                    {"in": "query", "name": "user_id", "required": True, "type": "integer"},
+                    {"in": "query", "name": "user_id", "required": True, "type": "string"},
                 ],
                 "responses": {
                     "200": {
@@ -280,7 +282,7 @@ def _make_openapi_2_schema(endpoints: Tuple[str, ...]) -> Dict:
                             "UpdateUserById": {
                                 "operationRef": "#/paths/~1users~1{user_id}/patch",
                                 "parameters": {"user_id": "$response.body#/id"},
-                                "requestBody": {"username": "foo"},
+                                "requestBody": {"first_name": "foo", "last_name": "bar"},
                             }
                         },
                     },
@@ -290,7 +292,7 @@ def _make_openapi_2_schema(endpoints: Tuple[str, ...]) -> Dict:
         elif endpoint == "update_user":
             parent = template["paths"].setdefault(path, {})
             parent["parameters"] = [
-                {"in": "path", "name": "user_id", "required": True, "type": "integer"},
+                {"in": "path", "name": "user_id", "required": True, "type": "string"},
                 {"in": "query", "name": "common", "required": True, "type": "integer"},
             ]
             schema = {
@@ -301,10 +303,14 @@ def _make_openapi_2_schema(endpoints: Tuple[str, ...]) -> Dict:
                         "name": "username",
                         "required": True,
                         "schema": {
-                            "additionalProperties": False,
                             "type": "object",
-                            "properties": {"username": {"type": "string"}},
-                            "required": ["username"],
+                            "properties": {
+                                "first_name": {"type": "string", "minLength": 3},
+                                # Note, the `last_name` field should not be nullable, it is a placed bug
+                                "last_name": {"type": "string", "minLength": 3, "x-nullable": True},
+                            },
+                            "required": ["first_name", "last_name"],
+                            "additionalProperties": False,
                         },
                     },
                 ],
@@ -525,8 +531,11 @@ def _make_openapi_3_schema(endpoints: Tuple[str, ...]) -> Dict:
                         "application/json": {
                             "schema": {
                                 "type": "object",
-                                "properties": {"username": {"type": "string", "minLength": 3}},
-                                "required": ["username"],
+                                "properties": {
+                                    "first_name": {"type": "string", "minLength": 3},
+                                    "last_name": {"type": "string", "minLength": 3},
+                                },
+                                "required": ["first_name", "last_name"],
                                 "additionalProperties": False,
                             }
                         }
@@ -540,7 +549,6 @@ def _make_openapi_3_schema(endpoints: Tuple[str, ...]) -> Dict:
                 {
                     "operationId": "updateUser",
                     "parameters": {"user_id": "$response.body#/id"},
-                    "requestBody": {"username": "foo"},
                 },
             )
             template["components"]["responses"] = {
@@ -560,12 +568,12 @@ def _make_openapi_3_schema(endpoints: Tuple[str, ...]) -> Dict:
             }
         elif endpoint == "get_user":
             parent = template["paths"].setdefault(path, {})
-            parent["parameters"] = [{"in": "path", "name": "user_id", "required": True, "schema": {"type": "integer"}}]
+            parent["parameters"] = [{"in": "path", "name": "user_id", "required": True, "schema": {"type": "string"}}]
             schema = {
                 "operationId": "getUser",
                 "parameters": [
                     {"in": "query", "name": "code", "required": True, "schema": {"type": "integer"}},
-                    {"in": "query", "name": "user_id", "required": True, "schema": {"type": "integer"}},
+                    {"in": "query", "name": "user_id", "required": True, "schema": {"type": "string"}},
                 ],
                 "responses": {
                     "200": {
@@ -574,7 +582,7 @@ def _make_openapi_3_schema(endpoints: Tuple[str, ...]) -> Dict:
                             "UpdateUserById": {
                                 "operationRef": "#/paths/~1users~1{user_id}/patch",
                                 "parameters": {"user_id": "$response.body#/id"},
-                                "requestBody": {"username": "foo"},
+                                "requestBody": {"first_name": "foo", "last_name": "bar"},
                             }
                         },
                     },
@@ -584,7 +592,7 @@ def _make_openapi_3_schema(endpoints: Tuple[str, ...]) -> Dict:
         elif endpoint == "update_user":
             parent = template["paths"].setdefault(path, {})
             parent["parameters"] = [
-                {"in": "path", "name": "user_id", "required": True, "schema": {"type": "integer"}},
+                {"in": "path", "name": "user_id", "required": True, "schema": {"type": "string"}},
                 {"in": "query", "name": "common", "required": True, "schema": {"type": "integer"}},
             ]
             schema = {
@@ -593,10 +601,14 @@ def _make_openapi_3_schema(endpoints: Tuple[str, ...]) -> Dict:
                     "content": {
                         "application/json": {
                             "schema": {
-                                "additionalProperties": False,
                                 "type": "object",
-                                "properties": {"username": {"type": "string"}},
-                                "required": ["username"],
+                                "properties": {
+                                    "first_name": {"type": "string", "minLength": 3},
+                                    # Note, the `last_name` field should not be nullable, it is a placed bug
+                                    "last_name": {"type": "string", "minLength": 3, "nullable": True},
+                                },
+                                "required": ["first_name", "last_name"],
+                                "additionalProperties": False,
                             }
                         }
                     },

--- a/test/specs/openapi/test_links.py
+++ b/test/specs/openapi/test_links.py
@@ -74,7 +74,6 @@ def response():
                     name="UpdateUserById",
                     endpoint=ANY,
                     parameters={"user_id": "$response.body#/id"},
-                    request_body={"username": "foo"},
                 ),
             ],
         ),
@@ -84,7 +83,7 @@ def response():
 @pytest.mark.endpoints("create_user", "get_user", "update_user")
 def test_get_links(openapi3_base_url, schema_url, url, expected):
     schema = schemathesis.from_uri(schema_url)
-    response = requests.post(f"{openapi3_base_url}{url}", json={"username": "TEST"})
+    response = requests.post(f"{openapi3_base_url}{url}", json={"first_name": "TEST", "last_name": "TEST"})
     assert schema.endpoints["/users/"]["POST"].get_stateful_tests(response, Stateful.links) == expected
 
 

--- a/test/specs/openapi/test_stateful.py
+++ b/test/specs/openapi/test_stateful.py
@@ -48,6 +48,44 @@ def test_default_status_code(response_status, status_codes, matching):
     assert filter_function(StepResult(response, None)) is matching
 
 
+def test_custom_rule(testdir, openapi3_base_url):
+    # When the state machine contains a failing rule that does not expect `Case`
+    testdir.make_test(
+        f"""
+from hypothesis.stateful import initialize, rule
+
+schema.base_url = "{openapi3_base_url}"
+
+class APIWorkflow(schema.as_state_machine()):
+
+    def validate_response(self, response, case):
+        pass
+
+    @rule(data=st.just("foo"))
+    def some(self, data):
+        assert 0
+
+TestStateful = APIWorkflow.TestCase
+""",
+    )
+    result = testdir.runpytest()
+    # Then the reproducing steps should be correctly displayed
+    result.assert_outcomes(failed=1)
+    result.stdout.re_match_lines([r"state.some\(data='foo'\)"])
+
+
+# With the following tests we try to uncover a bug that requires multiple steps with a shared step
+# There are three operations:
+#   1. Create user via first & last name
+#   2. Get user's details - returns id & full name
+#   3. Update user - update first & last name
+# The problem is that (3) allows you to update a user with non string last name, which is not detectable in its output.
+# However, (2) will fail when it will try to create a full name. Reproducing this bug requires 3 steps:
+#   1. Create a user
+#   2. Update the user with an invalid last name
+#   3. Get info about this user
+
+
 @pytest.mark.endpoints("create_user", "get_user", "update_user")
 def test_hidden_failure(testdir, app_schema, openapi3_base_url):
     # When we run test as a state machine
@@ -56,21 +94,20 @@ def test_hidden_failure(testdir, app_schema, openapi3_base_url):
 schema.base_url = "{openapi3_base_url}"
 TestStateful = schema.as_state_machine().TestCase
 TestStateful.settings = settings(
-    max_examples=300,
+    max_examples=200,
     deadline=None,
     derandomize=True,
     suppress_health_check=HealthCheck.all(),
-    stateful_step_count=5  # There is no need for longer sequences to uncover the bug
+    stateful_step_count=3  # There is no need for longer sequences to uncover the bug
 )
 """,
         schema=app_schema,
     )
     result = testdir.runpytest("--hypothesis-seed=42")
-    # Then it should be able to find a hidden error that happens on the following sequence of API calls:
-    # ["POST", "GET", "PATCH", "GET", "PATCH"]
+    # Then it should be able to find a hidden error:
     result.assert_outcomes(failed=1)
-    # And there should be Python code to reproduce the error in the PATCH call
-    result.stdout.re_match_lines([rf"E +requests\.patch\('{openapi3_base_url}/users/\d+.+"])
+    # And there should be Python code to reproduce the error in the GET call
+    result.stdout.re_match_lines([rf"E +requests\.get\('{openapi3_base_url}/users/\w+.+"])
     # And the reproducing example should work
     first = result.outlines.index("Falsifying example:") + 1
     last = result.outlines.index("state.teardown()") + 1
@@ -108,14 +145,13 @@ def test_hidden_failure_app(request, factory_name, open_api_3):
             target=schema["/users/{user_id}"]["PATCH"],
             status_code="201",
             parameters={"user_id": "$response.body#/id"},
-            request_body={"username": "foo"},
         )
         schema.add_link(
             source=schema["/users/{user_id}"]["GET"],
             target="#/paths/~1users~1{user_id}/patch",
             status_code="200",
             parameters={"user_id": "$response.body#/id"},
-            request_body={"username": "foo"},
+            request_body={"first_name": "foo", "last_name": "bar"},
         )
     else:
         schema = schemathesis.from_wsgi("/schema.yaml", app=app)
@@ -126,36 +162,10 @@ def test_hidden_failure_app(request, factory_name, open_api_3):
         run_state_machine_as_test(
             state_machine,
             settings=settings(
-                max_examples=300,
+                max_examples=500,
                 deadline=None,
                 suppress_health_check=HealthCheck.all(),
                 derandomize=True,
-                stateful_step_count=5,
+                stateful_step_count=3,
             ),
         )
-
-
-def test_custom_rule(testdir, openapi3_base_url):
-    # When the state machine contains a failing rule that does not expect `Case`
-    testdir.make_test(
-        f"""
-from hypothesis.stateful import initialize, rule
-
-schema.base_url = "{openapi3_base_url}"
-
-class APIWorkflow(schema.as_state_machine()):
-
-    def validate_response(self, response, case):
-        pass
-
-    @rule(data=st.just("foo"))
-    def some(self, data):
-        assert 0
-
-TestStateful = APIWorkflow.TestCase
-""",
-    )
-    result = testdir.runpytest()
-    # Then the reproducing steps should be correctly displayed
-    result.assert_outcomes(failed=1)
-    result.stdout.re_match_lines([r"state.some\(data='foo'\)"])

--- a/test/specs/openapi/test_stateful.py
+++ b/test/specs/openapi/test_stateful.py
@@ -1,4 +1,5 @@
 import pytest
+import requests
 from hypothesis import HealthCheck, settings
 from hypothesis.stateful import run_state_machine_as_test
 from requests import Response
@@ -169,3 +170,44 @@ def test_hidden_failure_app(request, factory_name, open_api_3):
                 stateful_step_count=3,
             ),
         )
+
+
+def test_all_operations_with_links(openapi3_base_url):
+    # See GH-965
+    # When all API operations have inbound links (one recursive in this exact case)
+    raw_schema = {
+        "openapi": "3.0.2",
+        "info": {"title": "Test", "description": "Test", "version": "0.1.0"},
+        "paths": {
+            "/payload": {
+                "post": {
+                    "operationId": "payload",
+                    "requestBody": {
+                        "content": {"application/json": {"schema": {"type": "integer", "minimum": 0, "maximum": 5}}},
+                        "required": True,
+                    },
+                    "responses": {
+                        "200": {
+                            "description": "OK",
+                            "links": {"More": {"operationId": "payload", "requestBody": "$response.body#/value"}},
+                        }
+                    },
+                }
+            }
+        },
+    }
+    schema = schemathesis.from_dict(raw_schema, base_url=openapi3_base_url)
+
+    class APIWorkflow(schema.as_state_machine()):
+        def call(self, case, **kwargs):
+            # The actual response doesn't matter much for this test
+            response = requests.Response()
+            response.status_code = 200
+            response._content = b'{"value": 42}'
+            return response
+
+    # Then test should be successful
+    # And there should be no "Unsatisfiable" error
+    run_state_machine_as_test(
+        APIWorkflow, settings=settings(max_examples=1, deadline=None, suppress_health_check=HealthCheck.all())
+    )

